packages/devtools_app/lib/src/performance/performance_controller.dart

```diff
@@ -248,56 +353,112 @@ class PerformanceController
+      _maybeAddUnassignedEventsToFrame(frame);
+    }
+    if (frame.isWellFormed) {
+      _updateFirstWellFormedFrameMicros(frame);
+    } else {
+      unassignedFlutterFrames[frame.id] = frame;
+    }
+  }
 
-    if (allTraceEvents.isEmpty) {
-      _emptyTimeline.value = true;
-      return;
+  void _maybeAddUnassignedEventsToFrame(FlutterFrame frame) {
+    _maybeAddUnassignedEventToFrame(frame, TimelineEventType.ui);
+    _maybeAddUnassignedEventToFrame(frame, TimelineEventType.raster);
+    if (unassignedFlutterFrameEvents[frame.id][TimelineEventType.ui.index] ==
+            null &&
+        unassignedFlutterFrameEvents[frame.id]
+                [TimelineEventType.raster.index] ==
+            null) {
+      unassignedFlutterFrameEvents.remove(frame.id);
     }
+  }
 
-    _processing.value = true;
-    await processTraceEvents(allTraceEvents);
-    _processing.value = false;
+  void _maybeAddUnassignedEventToFrame(
+    FlutterFrame frame,
+    TimelineEventType type,
+  ) {
+    final event = unassignedFlutterFrameEvents[frame.id][type.index];
+    if (event != null) {
+      frame.setEventFlow(event, type: type);
+      unassignedFlutterFrameEvents[frame.id][type.index] = null;
+    }
+  }
 
-    _flutterFrames.value = data.frames;
+  void _maybeAddEventToUnassignedFrame(
+    int frameNumber,
+    TimelineEvent event,
+    TimelineEventType type,
+  ) {
+    if (frameNumber != null) {
+      if (unassignedFlutterFrames.containsKey(frameNumber)) {
+        final frame = unassignedFlutterFrames[frameNumber];
+        frame.setEventFlow(event, type: type);
+        if (frame.isWellFormed) {
+          unassignedFlutterFrames.remove(frameNumber);
+          _updateFirstWellFormedFrameMicros(frame);
+        }
+      } else {
+        final unassignedEventsForFrame =
+            unassignedFlutterFrameEvents.putIfAbsent(
+          frameNumber,
+          () => List.generate(2, (_) => null),
+        );
+        unassignedEventsForFrame[event.type.index] = event;
+      }
+    }
+  }
 
-    _maybeBadgeTabForJankyFrames();
+  void _addPendingFlutterFrames() {
+    _pendingFlutterFrames.forEach(_maybeBadgeTabForJankyFrame);
+    data.frames.addAll(_pendingFlutterFrames);
+    _flutterFrames.addAll(_pendingFlutterFrames);
+    _pendingFlutterFrames.clear();
   }
 
-  void _maybeBadgeTabForJankyFrames() {
+  void _maybeBadgeTabForJankyFrame(FlutterFrame frame) {
     if (_badgeTabForJankyFrames.value) {
-      for (final frame in _flutterFrames.value) {
-        if (frame.isJanky(_displayRefreshRate.value)) {
-          serviceManager.errorBadgeManager
-              .incrementBadgeCount(PerformanceScreen.id);
-        }
+      if (frame.isJanky(_displayRefreshRate.value)) {
+        serviceManager.errorBadgeManager
+            .incrementBadgeCount(PerformanceScreen.id);
       }
     }
   }
 
+  void toggleRecordingFrames(bool recording) {
+    _recordingFrames.value = recording;
+    if (_recordingFrames.value) {
```

Comment:
tiny nit: maybe condition on `recording`  instead of the value of _recordingFrames.value